### DISCUSSION TITLE
Video stream interruption

### DISCUSSION THREAD

**Discussion by uniqueandriy** [uNetworking/uWebSockets.js#965] [Q&A]
Recently I ran into a bit of a puzzle while working on one of my projects. I've been working on adding a typical file streaming (with ranges support, cache control, etc.), and decided to test it out with a short 5-minute video. The initial playback was just fine, but I noticed Chrome was repeatedly reconnecting to get data ranges from the server.

After some debugging, I discovered that the connection was getting aborted, and it seemed like the browser wasn't the one causing it. Chrome kept thinking the connection was fine even when the server called the `.onAborted` callback, leading to the `net::ERR CONTENT LENGTH MISMATCH` error in the console. As a result, sometimes playback was interrupted altogether.

To make sure I hadn't messed up something, I decided to compare with the [video streaming example](https://github.com/uNetworking/uWebSockets.js/blob/master/examples/VideoStreamer.js), to which I added corking and a Content-Length header. The interruptions persisted, here're the logs:
```
Video size is: 69856219 bytes
Listening to port 4000
Stream was opened, openStreams: 1
Stream was closed, openStreams: 0
1: 30.776ms
Stream was opened, openStreams: 1
Stream was closed, openStreams: 0
2: 34.620s
Stream was opened, openStreams: 1
Stream was closed, openStreams: 0
3: 27.879s
Stream was opened, openStreams: 1
Stream was closed, openStreams: 0
4: 36.760s
Stream was opened, openStreams: 1
Stream was closed, openStreams: 0
5: 14.551s
```

Got me wondering, why is this happening? Typically, when you serve such a video from something like `nginx`, a single connection stays active until the stream ends. Even if there are rare reconnections, the browser figures out right away the connection is closed.

P.S.: uWebSockets.js v20.31.0, Node.js v18.12.1, Windows 10 Pro (Build 19041)

**Comment by kartikk221** [comment]
@uniqueandriy It sounds like you are performing incorrect backpressure handling.
See Issue: https://github.com/uNetworking/uWebSockets.js/issues/961 for the discussion.

With that said, If you wrote your code according to the video streaming example than this section of code may be problematic: https://github.com/uNetworking/uWebSockets.js/blob/ecd89ce24ae98fd3936e3387ccd315615d1eee37/examples/VideoStreamer.js#L41C7-L81

Essentially, this code consumes each chunk from the readStream with the `'data'` event which will gradually emit all chunks until done. As you are serving each of these chunks, some may not be sent due to backpressue in which case it implements the backpressure handling with `res.onWritable()`. The problem is that you cannot bind `res.onWritable()` more than one time throughout the whole lifecycle of a request. So what's likely happening is that the first time a chunk encounters backpressure, It retries successfully. The second time a chunk encounters backpressure, It is unable to properly retry the second chunk as the second `res.onWritable()` handler being assigned will not work properly.

As discussed in the issue above, I implemented a fix for this in [hyper-express](https://github.com/kartikk221/hyper-express) with proper backpressure handling recently, so feel free to put together a very quick demo to stream your large video file using hyper-express to see If that may resolve your issue.

**Reply by uniqueandriy** [reply]
Hello @kartikk221, thank you for your reply.

Yes, I actually thought that the `onWritable` would work like an event listener called once the backpressure is removed. For some reason didn't notice the issue while searching, although it is quite recent. However, after implementing the fix and binding once per response, nothing really changed much. Occasionally, the connection seemed to stay alive a bit longer, but that was about it. I added some logs to investigate further, but couldn't spot any obvious issues.

So, I decided to follow your advice and tried a comparison. I created a really simple file streaming server with `hyper-express` using the examples. Strangely enough, it turned out even less stable – the connection gets interrupted way more often, and sometimes they don't even last 100 milliseconds. Worth noting that the browser appears to detect interrupted connections much faster though.

Here's the code:
```
const fs = require('fs');
const HyperExpress = require('hyper-express');

const server = new HyperExpress.Server();

server.get('/', (req, res) => {
    res.send('Home');
});

server.get('/static/vid.mp4', (req, res) => {
   const fileStream = fs.createReadStream('./vid.mp4');
   fileStream.on('error', (err) => console.error(err));

   res
	.header('content-length', '69856219')
	.header('content-type', 'video/mp4')
	.stream(fileStream);

    // fileStream.pipe(res); // works the same
});


server.listen(4000)
    .then((socket) => console.log('Started on port 4000'))
    .catch((error) => console.log('Failed to start on port 4000'));
```

P.S.: hyper-express v6.8.5

**Reply by kartikk221** [reply]
- Could you specify the `content-length` as the second argument of the `.stream()` call? Currently you are specifying the content-length yourself through the header but `hyper-express` will still stream this with chunked-encoding as the `stream()` method requires you to provide the `total_size` as the second argument to do controlled writes.
- Could you try to listen for the Response events in hyper-express to see what the Response is doing as these connection interruptions occur?

**Reply by uniqueandriy** [reply]
Now the behavior is almost identical to what I observed in my project after adding the fix. The `abort` event is triggered.

**Reply by kartikk221** [reply]
I see, yeah in that case this would be a problem with `uWebsockets.js` as the `abort` event is triggered in a direct link to the `.onAborted()` which hyper-express has zero control over. So either the browser or uWS is disconnecting for some reason. I would recommend you make an issue for this so it can get a closer look.

Btw how long does it take before the `abort` event is triggered?

**Reply by uniqueandriy** [reply]
I suspected so, but decided to ask the question first, in case I missed something. Would probably make sense to create an issue out of this.

Typically, the interruption happens after about a minute (or a minute and a half) on average.

**Reply by porsager** [reply]
There is no problem calling onWritable multiple times - it's just a setter (see sample below)... I think the issue you faced @kartikk221 was using it like an event emitter, so anything but the last function you passed to onWritable wouldn't be called.

It's probably a bad idea to debug an issue by putting another layer of abstraction on top of the issue. Better cut away as much as possible.

Here is a video streaming sample with backpressure support using manual reads from a handle thereby also bypassing a bit more of nodes slowness.

```js
const uWS = require('uWebSockets.js')
const fsp = require('fs/promises')

const port = 9001

uWS.App().get('/*', async(r) => {
  let lastOffset = 0
    , read = 0
    , highWaterMark = 512 * 1024
    , buffer = Buffer.allocUnsafe(highWaterMark)
    , handle
    , aborted
    , abort

  r.onAborted(() => {
    aborted = true
    abort && abort()
  })

  handle = await fsp.open('some-video.mp4')

  r.cork(() => r.writeHeader('content-type', 'video/mp4'))
  const { size } = await handle.stat()

  while (!aborted && read < size) {
    const { bytesRead } = await handle.read(buffer, 0, Math.min(highWaterMark, size - read), read)
    if (aborted)
      break
    read += bytesRead
    lastOffset = r.getWriteOffset()
    let ok
    r.cork(() => [ok] = r.tryEnd(buffer.subarray(0, bytesRead), size))
    ok || await tryRest()
  }
  handle.close()

  function tryRest() {
    return new Promise(resolve => {
      abort = resolve
      r.onWritable(offset => {
        const [ok] = r.tryEnd(buffer.subarray(offset - lastOffset, bytesRead), size)
        ok && resolve()
        return ok
      })
    })
  }

}).listen(port, (token) => {
  token
    ? console.log('Listening to port ' + port)
    : console.log('Failed to listen to port ' + port)
})
```

**Reply by uniqueandriy** [reply]
Hi @porsager, thanks for your reply.

If you take a closer look, Kartik's code added to the issue seems to do pretty much the same thing as your sample, just in a more "indirect way". The `_stream_chunk` method returns a `Promise` that gets resolved when the chunk is successfully sent. To be honest, I can't think of a scenario where the handler could be rewritten.

In my project, setting the `onWritable` handler works a bit differently. It's inside an async function as well, but my class that handles the response extends the `Writable` stream (with a custom `_write` method).

I totally get what you mean about abstraction levels during debuging, but that wasn't my intention either. Even before I asked the question, I managed to reproduce the behavior using simple code based on the video streaming example (with a few small tweaks). I just wanted to see how it worked in an open source project, where the author recently ran into a similar issue, apparently.

And yes, I tried running the sample you provided and the behavior is the same. Although there were a few instances when the connection was not aborted for about two minutes.

**Reply by porsager** [reply]
Cool. Would like to be able to repro the issue, to look closer. How did you repro with the sample i sent?

**Reply by uniqueandriy** [reply]
1. **Prepare the Environment:**
   - Place the latest release of `uWebsockets.js`, a file with your sample, and a short video named `some-video.mp4` in a separate folder.

2. **Modify the Sample Code:**
   - Add two console logs to the sample code:
     - Place one at the beginning of the GET handler.
     - Add another one to track the connection abortion.

3. **Run the Script:**
   - Open a PowerShell window and run the sample script.

4. **Open Google Chrome:**
   - Open Google Chrome and navigate to `http://localhost:9001/`.

5. **Observe the First Connection:**
   - Note that the first connection is aborted almost immediately, which is normal behavior AFAIK. Keep an eye on the second connection.

6. **Wait for Aborted Console Log:**
   - Wait for approximately 30 seconds. You should observe the `aborted` console log in the PowerShell window. However, the browser will still think the connection is active.

7. **Observe Browser Behavior:**
   - Continue monitoring the browser. After about another 30 seconds, the browser itself closes the connection.

8. **Error Log in Browser Console:**
   - Check the browser console for an error message. You will see a log with the `net::ERR_CONTENT_LENGTH_MISMATCH 200 (OK)` error.

9. **Browser Starts a New Connection:**
   - Observe that the browser starts a new connection after encountering the error.

**Reply by porsager** [reply]
I don't see any issues on macos with my sample posted here and your instructions, and I haven't seen any on my linux server where I'm hosting VOD content this way (although with byte-range support too) :-) I thought you were simulating excessively slow connections to test backpressure, but this seems very straight forward.

I'll give windows a shot now.

**Reply by porsager** [reply]
Not seeing anything on windows 11 either. Sure you tested my sample verbatim? I've gone ahead and added some better handling of the file handle now.

**Reply by uasan** [reply]
@porsager I would place your example in the examples folder, I like it, high-quality and low-level programming

**Reply by uniqueandriy** [reply]
I literally copy-pasted the sample and added two console logs, and that's it. Will try to reproduce it on another system.

P.S.: I think you forgot to remove the last argument to the `tryRest` function call (`aborter`).

**Reply by porsager** [reply]
Interesting.

Yeah, I'm not denying the issue at all, I'm very curious since I'd obviously like to catch it in my own setup if there's something to it.

About tryRest (updated), we can just remove all the arguments now that it's in scope ;)

**Reply by uasan** [reply]
perhaps the error only appears using SSL, you tested without SSL or vice versa

**Reply by porsager** [reply]
That's a good point@uasan! I also have another issue myself with chunked transfers and SSL. Just need to find the time to describe and make a reproducible case.

**Reply by uniqueandriy** [reply]
Decided to try out the sample on a Fedora machine, which happens to have an older version of Node.js installed - v16.18.1. And, despite the variations in platform and module versions, the behavior remains the same.

Just to add: all of the tests were conducted on a server without SSL, on the same device and without any deliberate interference with the connections.

**Reply by uasan** [reply]
Yes, I reproduced this, for some reason the connection closes, in Chrome, in other browsers everything is fine
